### DISCUSSION TITLE
creating extensions

### DISCUSSION THREAD

**Discussion by lordtgm** [Minestom/Minestom#1820] [Help]
so I have an extension. check and tell me if I'm doing anything wrong
build.gradle:
```
plugins {
    id 'java'
    id 'maven-publish'
}

sourceCompatibility = 17

group 'com.github.lordtgm'
version '1.0'

repositories {
    mavenCentral()
    maven { url 'https://jitpack.io' }
    maven { url 'https://repo.minestom.net/repository/maven-public/' }
}

java {
    withJavadocJar()
    withSourcesJar()
}

dependencies {
    compileOnly 'com.github.Minestom.Minestom:Minestom:-SNAPSHOT'
    testImplementation 'com.github.Minestom.Minestom:Minestom:-SNAPSHOT'
    implementation group: 'net.java.dev.jna', name: 'jna', version: '5.13.0'
}

```
extension.json:
```
{
  "entrypoint": "com.github.lordtgm.minestom.TpsFixExtension",
  "name": "TpsFixExtension",
  "version": "1.0.0",
  "externalDependencies": {
    "repositories": [
      {"name": "Central", "url": "https://repo.minestom.net/repository/maven-public/"}
    ],
    "artifacts": [
      "net.java.dev.jna:5.13.0"
    ]
  }
}
```
then I used gradle task build and got 3 jar files in lib. what should I do know? how would I use my extension? just drop the main jar file in extensions folder?

**Comment by ZakShearman** [comment]
For your usecase, you probably don't need the sources or javadocs JAR so you can just remove the `withJavadocsJar`... removing these two will just generate you one JAR.

Additionally the repo `https://repo.minestom.net/repository/maven-public/` doesn't exist.

You'll also need to use Shadow to shade JNA in your extension:
```gradle
plugins {
    id("com.github.johnrengelman.shadow") version "7.1.2"
}
```
OR, use the external dependencies in the `extension.json` only and swap it from being `implementation` to `compileOnly`. If you do this, make sure to fix the repository URL.

After this then yes, you just put it in the extensions folder.

**Reply by lordtgm** [reply]
so should I remove that repo from both build.gradle and extension.json? if yes, what should I use in extensions.json?
